File: C++/1218. Longest Arithmetic Subsequence of Given Difference.cpp
class Solution {
public:
    int longestSubsequence(vector<int>& arr, int difference) {
        int m = 0, dp[40001] = {0};

        for (int i: arr) {
            i += 20001;
            dp[i] = dp[i - difference] != 0 ? dp[i - difference] + 1 : 1;
            m = max(m, dp[i]);
        }

        return m;
    }
};

File: Python/1218. Longest Arithmetic Subsequence of Given Difference.py
class Solution:
    def longestSubsequence(self, arr: List[int], difference: int) -> int:
        dp = {}

        for i in arr:
            if i - difference in dp: dp[i] = dp[i - difference] + 1
            else: dp[i] = 1

        return max(dp.values())
